File: js/test/ModCDPClient_protocol_validation.test.ts
// MODCDP_TRANSLATE_TEST: KEEP THIS TEST FILE TRANSLATED ACROSS TYPESCRIPT, PYTHON, AND GO.
// All test cases, descriptions, covered edge cases, and setup should be kept perfectly 1:1 in sync between:
// - ./python/tests/test_ModCDPClient_protocol_validation.py
// - ./go/modcdp/client/ModCDPClient_protocol_validation_test.go
// NO MOCKING, NO MONKEY PATCHING, NO SIMULATING, NO FAKING, NO SKIPPING ALLOWED.
// USE REAL USER-FACING CODE PATHS WITH REAL BROWSERS, REAL CLASSES, REAL URLS, etc. Hard fail if keys or other env requirements are missing.
import assert from "node:assert/strict";
import { test } from "vitest";
import { z } from "zod";

import { ModCDPClient } from "../src/index.js";
import { CDPTypes } from "../src/types/CDPTypes.js";
import type { cdp } from "../src/types/generated/cdp.js";

test("native CDP schemas validate method params, return values, and event payloads statically and at runtime", () => {
  const types = new CDPTypes();
  const client = new ModCDPClient({
    launcher: { launcher_mode: "none" },
    upstream: { upstream_mode: "ws" },
    injector: { injector_mode: "none" },
    server_config: null,
  });
  const runtime_params: cdp.types.ts.Runtime.EvaluateParams = {
    expression: "1 + 1",
    returnByValue: true,
  };
  const runtime_result: cdp.types.ts.Runtime.EvaluateResult = {
    result: { type: "number", value: 2, description: "2" },
  };
  const target_event: cdp.types.ts.Target.TargetCreatedEvent = {
    targetInfo: {
      targetId: "target-1",
      type: "page",
      title: "Example",
      url: "https://example.com",
      attached: false,
      canAccessOpener: false,
    },
  };

  if (false) {
    const alias_result: Awaited<ReturnType<typeof client.Runtime.evaluate>> = runtime_result;
    void alias_result;
    client.on(client.Target.targetCreated, (event) => {
      const targetId: string = event.targetInfo.targetId;
      void targetId;
      // @ts-expect-error targetId is a string, not a number.
      const badTargetId: number = event.targetInfo.targetId;
      void badTargetId;
    });
    // @ts-expect-error Runtime.evaluate params require expression.
    client.Runtime.evaluate({ returnByValue: true });
    const badRuntimeResult: cdp.types.ts.Runtime.EvaluateResult = {
      // @ts-expect-error Runtime.evaluate result must be a Runtime.RemoteObject.
      result: { type: 1 },
    };
    void badRuntimeResult;
    const badTargetEvent: cdp.types.ts.Target.TargetCreatedEvent = {
      targetInfo: {
        ...target_event.targetInfo,
        // @ts-expect-error Target.targetCreated targetInfo.targetId is a string.
        targetId: 1,
      },
    };
    void badTargetEvent;
  }

  assert.deepEqual(types.parseCommandParams("Runtime.evaluate", runtime_params), runtime_params);
  assert.deepEqual(types.parseCommandResult("Runtime.evaluate", runtime_result), runtime_result);
  assert.deepEqual(types.parseEventPayload("Target.targetCreated", target_event), target_event);
  assert.throws(() => types.parseCommandParams("Runtime.evaluate", { returnByValue: true }));
  assert.throws(() => types.parseCommandResult("Runtime.evaluate", {}));
  assert.throws(() =>
    types.parseEventPayload("Target.targetCreated", {
      targetInfo: { ...target_event.targetInfo, targetId: 1 },
    }),
  );
});

test("Mod schemas validate method params, return values, event payloads, and middleware registrations statically and at runtime", () => {
  const types = new CDPTypes();
  const client = new ModCDPClient({
    launcher: { launcher_mode: "none" },
    upstream: { upstream_mode: "ws" },
    injector: { injector_mode: "none" },
    server_config: null,
  });
  const ping_params: cdp.types.ts.Mod.PingParams = { sent_at: 123 };
  const ping_result: cdp.types.ts.Mod.PingResponse = { ok: true };
  const pong_event: cdp.types.ts.Mod.PongEvent = {
    sent_at: 123,
    received_at: 124,
    from: "extension-service-worker",
  };
  const middleware_params: cdp.types.ts.Mod.AddMiddlewareParams = {
    name: client.Target.getTargets,
    phase: client.RESPONSE,
    expression: "async (payload, next) => next(payload)",
  };
  const middleware_result: cdp.types.ts.Mod.AddMiddlewareResponse = {
    name: "Target.getTargets",
    phase: "response",
    registered: true,
  };

  if (false) {
    const alias_result: Awaited<ReturnType<typeof client.Mod.ping>> = ping_result;
    void alias_result;
    const alias_middleware_result: Awaited<ReturnType<typeof client.Mod.addMiddleware>> = middleware_result;
    void alias_middleware_result;
    // @ts-expect-error Mod.ping sent_at must be a number.
    client.Mod.ping({ sent_at: "123" });
    client.Mod.addMiddleware({
      name: "Custom.any",
      // @ts-expect-error middleware phase must be request, response, or event.
      phase: "after",
      expression: "async (payload, next) => next(payload)",
    });
    // @ts-expect-error Mod.pong received_at is required.
    const badPongEvent: cdp.types.ts.Mod.PongEvent = {
      sent_at: 123,
      from: "extension-service-worker",
    };
    void badPongEvent;
  }

  assert.deepEqual(types.parseCommandParams("Mod.ping", ping_params), {
    sent_at: 123,
  });
  assert.deepEqual(types.parseCommandResult("Mod.ping", ping_result), {
    ok: true,
  });
  assert.deepEqual(types.parseEventPayload("Mod.pong", pong_event), pong_event);
  assert.deepEqual(types.parseCommandParams("Mod.addMiddleware", middleware_params), middleware_params);
  assert.deepEqual(types.parseCommandResult("Mod.addMiddleware", middleware_result), middleware_result);
  assert.throws(() => types.parseCommandParams("Mod.ping", { sent_at: "123" }));
  assert.throws(() => types.parseCommandResult("Mod.ping", { ok: "true" }));
  assert.throws(() =>
    types.parseEventPayload("Mod.pong", {
      sent_at: 123,
      from: "extension-service-worker",
    }),
  );
  assert.throws(() =>
    types.parseCommandParams("Mod.addMiddleware", {
      name: "Custom.any",
      phase: "after",
      expression: "async (payload, next) => next(payload)",
    }),
  );
  assert.throws(() =>
    types.parseCommandResult("Mod.addMiddleware", {
      name: "Custom.any",
      phase: "after",
      registered: true,
    }),
  );
});

test("constructor custom schemas validate command params, return values, events, and middleware registrations statically and at runtime", () => {
  const SumParams = z.object({ left: z.number(), right: z.number() });
  const SumResult = z.object({ value: z.number() });
  const FinishedEvent = z.object({ total: z.number(), label: z.string() });
  const client = new ModCDPClient({
    launcher: { launcher_mode: "none" },
    upstream: { upstream_mode: "ws" },
    injector: { injector_mode: "none" },
    server_config: null,
    types: {
      custom_commands: {
        "Custom.sum": {
          params_schema: SumParams,
          result_schema: SumResult,
          expression: "async ({ left, right }) => ({ value: left + right })",
        },
      },
      custom_events: {
        "Custom.finished": { event_schema: FinishedEvent },
      },
      custom_middlewares: [
        {
          name: "Custom.sum",
          phase: "response",
          expression: "async (payload, next) => next(payload)",
        },
      ],
    },
  });

  if (false) {
    const params: Parameters<typeof client.Custom.sum>[0] = {
      left: 1,
      right: 2,
    };
    void params;
    const result: Awaited<ReturnType<typeof client.Custom.sum>> = { value: 3 };
    void result;
    client.on("Custom.finished", (event) => {
      const total: number = event.total;
      void total;
      // @ts-expect-error Custom.finished total is a number.
      const badTotal: string = event.total;
      void badTotal;
    });
    // @ts-expect-error Custom.sum requires numeric left/right params.
    client.Custom.sum({ left: "1", right: 2 });
    // @ts-expect-error Custom.sum returns a CDP-style result object.
    const badResult: Awaited<ReturnType<typeof client.Custom.sum>> = 3;
    void badResult;
  }

  assert.deepEqual(client.types.parseCommandParams("Custom.sum", { left: 1, right: 2 }), { left: 1, right: 2 });
  assert.deepEqual(client.types.parseCommandResult("Custom.sum", { value: 3 }), { value: 3 });
  assert.deepEqual(
    client.types.parseEventPayload("Custom.finished", {
      total: 3,
      label: "ok",
    }),
    { total: 3, label: "ok" },
  );
  assert.throws(() => client.types.parseCommandParams("Custom.sum", { left: "1", right: 2 }));
  assert.throws(() => client.types.parseCommandResult("Custom.sum", { value: "3" }));
  assert.throws(() =>
    client.types.parseEventPayload("Custom.finished", {
      total: "3",
      label: "ok",
    }),
  );
  assert.deepEqual(client.types.customMiddlewareWireRegistrations(), [
    {
      name: "Custom.sum",
      phase: "response",
      expression: "async (payload, next) => next(payload)",
    },
  ]);
  assert.throws(
    () =>
      new CDPTypes({
        custom_middlewares: [
          {
            name: "Custom.sum",
            phase: "after",
            expression: "async (payload, next) => next(payload)",
          } as never,
        ],
      }),
  );
});

test("dynamic Mod registration updates custom command, event, and middleware validation", async () => {
  const client = new ModCDPClient({
    launcher: { launcher_mode: "none" },
    upstream: { upstream_mode: "ws" },
    injector: { injector_mode: "none" },
    server_config: null,
  });

  assert.deepEqual(
    await client.Mod.addCustomCommand("Custom.dynamic", {
      params_schema: z.object({ text: z.string().min(1) }),
      result_schema: z.object({ ok: z.boolean() }),
    }),
    { name: "Custom.dynamic", registered: true },
  );
  assert.deepEqual(
    await client.Mod.addCustomEvent("Custom.dynamicReady", {
      event_schema: z.object({ id: z.string().uuid() }),
    }),
    { name: "Custom.dynamicReady", registered: true },
  );
  assert.deepEqual(
    await client.Mod.addMiddleware({
      name: "Custom.dynamic",
      phase: client.RESPONSE,
      expression: "async (payload, next) => next(payload)",
    }),
    { name: "Custom.dynamic", phase: "response", registered: true },
  );

  assert.equal(typeof (client as unknown as { Custom: { dynamic: unknown } }).Custom.dynamic, "function");
  assert.deepEqual(client.types.parseCommandParams("Custom.dynamic", { text: "ok" }), { text: "ok" });
  assert.deepEqual(client.types.parseCommandResult("Custom.dynamic", { ok: true }), { ok: true });
  assert.deepEqual(
    client.types.parseEventPayload("Custom.dynamicReady", {
      id: "550e8400-e29b-41d4-a716-446655440000",
    }),
    { id: "550e8400-e29b-41d4-a716-446655440000" },
  );
  assert.deepEqual(client.types.customMiddlewareWireRegistrations(), [
    {
      name: "Custom.dynamic",
      phase: "response",
      expression: "async (payload, next) => next(payload)",
    },
  ]);
  assert.throws(() => client.types.parseCommandParams("Custom.dynamic", { text: "" }));
  assert.throws(() => client.types.parseCommandResult("Custom.dynamic", { ok: "yes" }));
  assert.throws(() => client.types.parseEventPayload("Custom.dynamicReady", { id: "nope" }));
  await assert.rejects(() =>
    client.Mod.addMiddleware({
      name: "Custom.dynamic",
      phase: "after" as never,
      expression: "async (payload, next) => next(payload)",
    }),
  );
});

test("client.types update replaces the registry with extended runtime validation and preserves static custom aliases on typed clients", () => {
  const client = new ModCDPClient({
    launcher: { launcher_mode: "none" },
    upstream: { upstream_mode: "ws" },
    injector: { injector_mode: "none" },
    server_config: null,
  });
  const updated_types = client.types.update({
    custom_commands: {
      "Custom.updated": {
        params_schema: z.object({ count: z.number().int().positive() }),
        result_schema: z.object({ done: z.boolean() }),
      },
    },
    custom_events: {
      "Custom.updatedReady": { event_schema: z.object({ ready: z.boolean() }) },
    },
    custom_middlewares: [
      {
        name: "Custom.updated",
        phase: "request",
        expression: "async (payload, next) => next(payload)",
      },
    ],
  });
  const typed_client = new ModCDPClient({
    launcher: { launcher_mode: "none" },
    upstream: { upstream_mode: "ws" },
    injector: { injector_mode: "none" },
    server_config: null,
    types: updated_types,
  });
  client.types = updated_types;

  if (false) {
    const params: Parameters<typeof typed_client.Custom.updated>[0] = {
      count: 1,
    };
    void params;
    const result: Awaited<ReturnType<typeof typed_client.Custom.updated>> = { done: true };
    void result;
    typed_client.on("Custom.updatedReady", (event) => {
      const ready: boolean = event.ready;
      void ready;
      // @ts-expect-error Custom.updatedReady.ready is boolean.
      const badReady: string = event.ready;
      void badReady;
    });
    // @ts-expect-error Custom.updated count is required.
    typed_client.Custom.updated({});
    // @ts-expect-error Custom.updated returns a CDP-style result object.
    const badResult: Awaited<ReturnType<typeof typed_client.Custom.updated>> = true;
    void badResult;
  }

  assert.equal(typeof (client as unknown as { Custom: { updated: unknown } }).Custom.updated, "function");
  assert.deepEqual(client.types.parseCommandParams("Custom.updated", { count: 1 }), { count: 1 });
  assert.deepEqual(client.types.parseCommandResult("Custom.updated", { done: true }), { done: true });
  assert.deepEqual(client.types.parseEventPayload("Custom.updatedReady", { ready: true }), { ready: true });
  assert.deepEqual(client.types.customMiddlewareWireRegistrations(), [
    {
      name: "Custom.updated",
      phase: "request",
      expression: "async (payload, next) => next(payload)",
    },
  ]);
  assert.throws(() => client.types.parseCommandParams("Custom.updated", { count: 0 }));
  assert.throws(() => client.types.parseCommandResult("Custom.updated", { done: "true" }));
  assert.throws(() => client.types.parseEventPayload("Custom.updatedReady", { ready: "true" }));
});

File: go/modcdp/client/ModCDPClient_protocol_validation_test.go
// MODCDP_TRANSLATE_TEST: KEEP THIS TEST FILE TRANSLATED ACROSS TYPESCRIPT, PYTHON, AND GO.
// All test cases, descriptions, covered edge cases, and setup should be kept perfectly 1:1 in sync between:
// - ./js/test/ModCDPClient_protocol_validation.test.ts
// - ./python/tests/test_ModCDPClient_protocol_validation.py
// NO MOCKING, NO MONKEY PATCHING, NO SIMULATING, NO FAKING, NO SKIPPING ALLOWED.
// USE REAL USER-FACING CODE PATHS WITH REAL BROWSERS, REAL CLASSES, REAL URLS, etc. Hard fail if keys or other env requirements are missing.
package client

import (
	"testing"

	abxjsonschema "github.com/ArchiveBox/abxbus/abxbus-go/v2/jsonschema"
)

type protocolValidationSumParams struct {
	Left  int `json:"left"`
	Right int `json:"right"`
}

type protocolValidationSumResult struct {
	Value int `json:"value"`
}

type protocolValidationFinishedEvent struct {
	Total int    `json:"total"`
	Label string `json:"label"`
}

type protocolValidationDynamicResult struct {
	OK bool `json:"ok"`
}

type protocolValidationUpdatedResult struct {
	Done bool `json:"done"`
}

type protocolValidationUpdatedReadyEvent struct {
	Ready bool `json:"ready"`
}

func TestNativeCDPSchemasValidateMethodParamsReturnValuesAndEventPayloadsStaticallyAndAtRuntime(t *testing.T) {
	types := NewCDPTypes(nil, nil, nil)
	cdp := New(Config{
		Launcher:     LauncherConfig{LauncherMode: "none"},
		Upstream:     UpstreamTransportConfig{UpstreamMode: "ws"},
		Injector:     InjectorConfig{InjectorMode: "none"},
		ServerConfig: ServerConfigNone,
	})
	runtimeParams := RuntimeEvaluateParams{Expression: "1 + 1", ReturnByValue: Bool(true)}
	runtimeResult := RuntimeEvaluateResult{Result: RuntimeRemoteObject{Type: "number", Value: 2, Description: String("2")}}
	targetEvent := TargetTargetCreatedEvent{
		TargetInfo: TargetTargetInfo{
			TargetID:        TargetTargetID("target-1"),
			Type:            "page",
			Title:           "Example",
			URL:             "https://example.com",
			Attached:        false,
			CanAccessOpener: false,
		},
	}

	_ = cdp.Runtime
	if _, err := types.ParseCommandParams("Runtime.evaluate", mustParamsMap(t, runtimeParams)); err != nil {
		t.Fatalf("Runtime.evaluate params should validate: %v", err)
	}
	if _, err := types.ParseCommandResult("Runtime.evaluate", runtimeResult); err != nil {
		t.Fatalf("Runtime.evaluate result should validate: %v", err)
	}
	if _, err := types.ParseEventPayload("Target.targetCreated", targetEvent); err != nil {
		t.Fatalf("Target.targetCreated event should validate: %v", err)
	}
	if _, err := types.ParseCommandParams("Runtime.evaluate", map[string]any{"returnByValue": true}); err == nil {
		t.Fatal("expected Runtime.evaluate params validation to reject missing expression")
	}
	if _, err := types.ParseCommandResult("Runtime.evaluate", map[string]any{}); err == nil {
		t.Fatal("expected Runtime.evaluate result validation to reject missing result")
	}
	if _, err := types.ParseEventPayload("Target.targetCreated", map[string]any{
		"targetInfo": map[string]any{
			"targetId":        1,
			"type":            "page",
			"title":           "Example",
			"url":             "https://example.com",
			"attached":        false,
			"canAccessOpener": false,
		},
	}); err == nil {
		t.Fatal("expected Target.targetCreated event validation to reject numeric targetId")
	}
}

func TestModSchemasValidateMethodParamsReturnValuesEventPayloadsAndMiddlewareRegistrationsStaticallyAndAtRuntime(t *testing.T) {
	types := NewCDPTypes(nil, nil, nil)
	pingParams := map[string]any{"sent_at": 123}
	pingResult := map[string]any{"ok": true}
	pongEvent := map[string]any{"sent_at": 123, "received_at": 124, "from": "extension-service-worker"}
	middlewareParams := map[string]any{
		"name":       "Target.getTargets",
		"phase":      "response",
		"expression": "async (payload, next) => next(payload)",
	}
	middlewareResult := map[string]any{"name": "Target.getTargets", "phase": "response", "registered": true}

	if parsed, err := types.ParseCommandParams("Mod.ping", pingParams); err != nil || parsed["sent_at"] != 123 {
		t.Fatalf("Mod.ping params = %#v, %v", parsed, err)
	}
	if parsed, err := types.ParseCommandResult("Mod.ping", pingResult); err != nil {
		t.Fatalf("Mod.ping result should validate: %#v, %v", parsed, err)
	}
	if parsed, err := types.ParseEventPayload("Mod.pong", pongEvent); err != nil || parsed == nil {
		t.Fatalf("Mod.pong event = %#v, %v", parsed, err)
	}
	if parsed, err := types.ParseCommandParams("Mod.addMiddleware", middlewareParams); err != nil || parsed["phase"] != "response" {
		t.Fatalf("Mod.addMiddleware params = %#v, %v", parsed, err)
	}
	if parsed, err := types.ParseCommandResult("Mod.addMiddleware", middlewareResult); err != nil {
		t.Fatalf("Mod.addMiddleware result should validate: %#v, %v", parsed, err)
	}
	if _, err := types.ParseCommandParams("Mod.ping", map[string]any{"sent_at": "123"}); err == nil {
		t.Fatal("expected Mod.ping params validation to reject string sent_at")
	}
	if _, err := types.ParseCommandResult("Mod.ping", map[string]any{"ok": "true"}); err == nil {
		t.Fatal("expected Mod.ping result validation to reject string ok")
	}
	if _, err := types.ParseEventPayload("Mod.pong", map[string]any{"sent_at": 123, "from": "extension-service-worker"}); err == nil {
		t.Fatal("expected Mod.pong event validation to reject missing received_at")
	}
	if _, err := types.ParseCommandParams("Mod.addMiddleware", map[string]any{"name": "Custom.any", "phase": "after", "expression": "async (payload, next) => next(payload)"}); err == nil {
		t.Fatal("expected Mod.addMiddleware params validation to reject invalid phase")
	}
	if _, err := types.ParseCommandResult("Mod.addMiddleware", map[string]any{"name": "Custom.any", "phase": "after", "registered": true}); err == nil {
		t.Fatal("expected Mod.addMiddleware result validation to reject invalid phase")
	}
}

func TestConstructorCustomSchemasValidateCommandParamsReturnValuesEventsAndMiddlewareRegistrationsStaticallyAndAtRuntime(t *testing.T) {
	cdp := New(Config{
		Launcher:     LauncherConfig{LauncherMode: "none"},
		Upstream:     UpstreamTransportConfig{UpstreamMode: "ws"},
		Injector:     InjectorConfig{InjectorMode: "none"},
		ServerConfig: ServerConfigNone,
		Types: &CDPTypesConfig{
			CustomCommands: []CustomCommand{{
				Name:         "Custom.sum",
				ParamsSchema: abxjsonschema.SchemaFor[protocolValidationSumParams](),
				ResultSchema: abxjsonschema.SchemaFor[protocolValidationSumResult](),
				Expression:   "async ({ left, right }) => ({ value: left + right })",
			}},
			CustomEvents: []CustomEvent{{
				Name:        "Custom.finished",
				EventSchema: abxjsonschema.SchemaFor[protocolValidationFinishedEvent](),
			}},
			CustomMiddlewares: []CustomMiddleware{{
				Name:       "Custom.sum",
				Phase:      "response",
				Expression: "async (payload, next) => next(payload)",
			}},
		},
	})

	if _, err := cdp.Types.ParseCommandParams("Custom.sum", map[string]any{"left": 1, "right": 2}); err != nil {
		t.Fatalf("Custom.sum params should validate: %v", err)
	}
	if _, err := cdp.Types.ParseCommandResult("Custom.sum", map[string]any{"value": 3}); err != nil {
		t.Fatalf("Custom.sum result should validate: %v", err)
	}
	if _, err := cdp.Types.ParseEventPayload("Custom.finished", map[string]any{"total": 3, "label": "ok"}); err != nil {
		t.Fatalf("Custom.finished event should validate: %v", err)
	}
	if _, err := cdp.Types.ParseCommandParams("Custom.sum", map[string]any{"left": "1", "right": 2}); err == nil {
		t.Fatal("expected Custom.sum params validation to reject string left")
	}
	if _, err := cdp.Types.ParseCommandResult("Custom.sum", map[string]any{"value": "3"}); err == nil {
		t.Fatal("expected Custom.sum result validation to reject string value")
	}
	if _, err := cdp.Types.ParseEventPayload("Custom.finished", map[string]any{"total": "3", "label": "ok"}); err == nil {
		t.Fatal("expected Custom.finished event validation to reject string total")
	}
	middlewares := cdp.Types.CustomMiddlewareWireRegistrations()
	if len(middlewares) != 1 || middlewares[0].Name != "Custom.sum" || middlewares[0].Phase != "response" || middlewares[0].Expression != "async (payload, next) => next(payload)" {
		t.Fatalf("custom middlewares = %#v", middlewares)
	}
	expectPanic(t, func() {
		NewCDPTypes(nil, nil, []CustomMiddleware{{Name: "Custom.sum", Phase: "after", Expression: "async (payload, next) => next(payload)"}})
	})
}

func TestDynamicModRegistrationUpdatesCustomCommandEventAndMiddlewareValidation(t *testing.T) {
	cdp := New(Config{
		Launcher:     LauncherConfig{LauncherMode: "none"},
		Upstream:     UpstreamTransportConfig{UpstreamMode: "ws"},
		Injector:     InjectorConfig{InjectorMode: "none"},
		ServerConfig: ServerConfigNone,
	})

	registered, err := cdp.Mod.AddCustomCommand(CustomCommand{
		Name: "Custom.dynamic",
		ParamsSchema: map[string]any{
			"type":                 "object",
			"properties":           map[string]any{"text": map[string]any{"type": "string", "minLength": 1}},
			"required":             []any{"text"},
			"additionalProperties": false,
		},
		ResultSchema: abxjsonschema.SchemaFor[protocolValidationDynamicResult](),
	})
	if err != nil {
		t.Fatal(err)
	}
	if registration, ok := registered.(map[string]any); !ok || registration["name"] != "Custom.dynamic" || registration["registered"] != true {
		t.Fatalf("Custom.dynamic registration = %#v", registered)
	}
	registered, err = cdp.Mod.AddCustomEvent(CustomEvent{
		Name: "Custom.dynamicReady",
		EventSchema: map[string]any{
			"type":                 "object",
			"properties":           map[string]any{"id": map[string]any{"type": "string", "pattern": "^[0-9a-f-]{36}$"}},
			"required":             []any{"id"},
			"additionalProperties": false,
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	if registration, ok := registered.(map[string]any); !ok || registration["name"] != "Custom.dynamicReady" || registration["registered"] != true {
		t.Fatalf("Custom.dynamicReady registration = %#v", registered)
	}
	registered, err = cdp.Mod.AddMiddleware(CustomMiddleware{
		Name:       "Custom.dynamic",
		Phase:      "response",
		Expression: "async (payload, next) => next(payload)",
	})
	if err != nil {
		t.Fatal(err)
	}
	if registration, ok := registered.(map[string]any); !ok || registration["name"] != "Custom.dynamic" || registration["phase"] != "response" || registration["registered"] != true {
		t.Fatalf("Custom.dynamic middleware registration = %#v", registered)
	}

	if _, ok := cdp.Types.CustomCommands["Custom.dynamic"]; !ok {
		t.Fatal("expected Custom.dynamic registry entry")
	}
	if _, err := cdp.Types.ParseCommandParams("Custom.dynamic", map[string]any{"text": "ok"}); err != nil {
		t.Fatalf("Custom.dynamic params should validate: %v", err)
	}
	if _, err := cdp.Types.ParseCommandResult("Custom.dynamic", map[string]any{"ok": true}); err != nil {
		t.Fatalf("Custom.dynamic result should validate: %v", err)
	}
	if _, err := cdp.Types.ParseEventPayload("Custom.dynamicReady", map[string]any{"id": "550e8400-e29b-41d4-a716-446655440000"}); err != nil {
		t.Fatalf("Custom.dynamicReady event should validate: %v", err)
	}
	middlewares := cdp.Types.CustomMiddlewareWireRegistrations()
	if len(middlewares) != 1 || middlewares[0].Name != "Custom.dynamic" || middlewares[0].Phase != "response" {
		t.Fatalf("custom middlewares = %#v", middlewares)
	}
	if _, err := cdp.Types.ParseCommandParams("Custom.dynamic", map[string]any{"text": ""}); err == nil {
		t.Fatal("expected Custom.dynamic params validation to reject empty text")
	}
	if _, err := cdp.Types.ParseCommandResult("Custom.dynamic", map[string]any{"ok": "yes"}); err == nil {
		t.Fatal("expected Custom.dynamic result validation to reject string ok")
	}
	if _, err := cdp.Types.ParseEventPayload("Custom.dynamicReady", map[string]any{"id": "nope"}); err == nil {
		t.Fatal("expected Custom.dynamicReady event validation to reject invalid uuid")
	}
	if _, err := cdp.Mod.AddMiddleware(CustomMiddleware{Name: "Custom.dynamic", Phase: "after", Expression: "async (payload, next) => next(payload)"}); err == nil {
		t.Fatal("expected invalid middleware phase to fail")
	}
}

func TestClientTypesUpdateReplacesTheRegistryWithExtendedRuntimeValidationAndPreservesStaticCustomAliasesOnTypedClients(t *testing.T) {
	cdp := New(Config{
		Launcher:     LauncherConfig{LauncherMode: "none"},
		Upstream:     UpstreamTransportConfig{UpstreamMode: "ws"},
		Injector:     InjectorConfig{InjectorMode: "none"},
		ServerConfig: ServerConfigNone,
	})
	updatedTypes := cdp.Types.Update(CDPTypesConfig{
		CustomCommands: []CustomCommand{{
			Name: "Custom.updated",
			ParamsSchema: map[string]any{
				"type":                 "object",
				"properties":           map[string]any{"count": map[string]any{"type": "integer", "minimum": 1}},
				"required":             []any{"count"},
				"additionalProperties": false,
			},
			ResultSchema: abxjsonschema.SchemaFor[protocolValidationUpdatedResult](),
		}},
		CustomEvents: []CustomEvent{{
			Name:        "Custom.updatedReady",
			EventSchema: abxjsonschema.SchemaFor[protocolValidationUpdatedReadyEvent](),
		}},
		CustomMiddlewares: []CustomMiddleware{{
			Name:       "Custom.updated",
			Phase:      "request",
			Expression: "async (payload, next) => next(payload)",
		}},
	})
	typedClient := New(Config{
		Launcher:     LauncherConfig{LauncherMode: "none"},
		Upstream:     UpstreamTransportConfig{UpstreamMode: "ws"},
		Injector:     InjectorConfig{InjectorMode: "none"},
		ServerConfig: ServerConfigNone,
	})
	typedClient.Types = updatedTypes
	cdp.Types = updatedTypes

	if _, ok := typedClient.Types.CustomCommands["Custom.updated"]; !ok {
		t.Fatal("expected typed client Custom.updated registry entry")
	}
	if _, ok := cdp.Types.CustomCommands["Custom.updated"]; !ok {
		t.Fatal("expected client Custom.updated registry entry")
	}
	if _, err := cdp.Types.ParseCommandParams("Custom.updated", map[string]any{"count": 1}); err != nil {
		t.Fatalf("Custom.updated params should validate: %v", err)
	}
	if _, err := cdp.Types.ParseCommandResult("Custom.updated", map[string]any{"done": true}); err != nil {
		t.Fatalf("Custom.updated result should validate: %v", err)
	}
	if _, err := cdp.Types.ParseEventPayload("Custom.updatedReady", map[string]any{"ready": true}); err != nil {
		t.Fatalf("Custom.updatedReady event should validate: %v", err)
	}
	middlewares := cdp.Types.CustomMiddlewareWireRegistrations()
	if len(middlewares) != 1 || middlewares[0].Name != "Custom.updated" || middlewares[0].Phase != "request" {
		t.Fatalf("custom middlewares = %#v", middlewares)
	}
	if _, err := cdp.Types.ParseCommandParams("Custom.updated", map[string]any{"count": 0}); err == nil {
		t.Fatal("expected Custom.updated params validation to reject zero count")
	}
	if _, err := cdp.Types.ParseCommandResult("Custom.updated", map[string]any{"done": "true"}); err == nil {
		t.Fatal("expected Custom.updated result validation to reject string done")
	}
	if _, err := cdp.Types.ParseEventPayload("Custom.updatedReady", map[string]any{"ready": "true"}); err == nil {
		t.Fatal("expected Custom.updatedReady event validation to reject string ready")
	}
}

func mustParamsMap(t *testing.T, value any) map[string]any {
	t.Helper()
	params, err := cdpParamsMap(value)
	if err != nil {
		t.Fatal(err)
	}
	return params
}
